### PR TITLE
Model Classifier

### PR DESCRIPTION
**Pull request by Appearls**
Created a Classifier for language detection.
Trained the model on 5 languages: Hindi, Telugu, Malayalam, Marathi, Bengali
Created a hindi translation model with Sequence to Sequence network and LSTM(Long Short Term Memory) to encode and decode texts with attention. 
databn.tsv: Dakshina  dataset used to train for Bengali
datahi.tsv: Dakshina  dataset used to train for HIndi
dataml.tsv: Dakshina  dataset used to train for Malayalam
datamr.tsv: Dakshina  dataset used to train for Marathi
datate.tsv: Dakshina  dataset used to train for Telugu
Classifier.py: Code used to create , train and test the model
language_classifier.h5:Classifier model in HDF5 format 
language_classifier.keras: Classifier model in keras format
preprocess.py: Code for processeing data to make it trainable
langhi.py: Code to train and test a model to translate from romanized hindi to native devanagari script
hindi_transliteration_model.keras: Hindi Translation model in keras format
hindi_transliteration_model.h5: Hindi Translation model in HDF5  format
langte.py: Code to train and test a model to translate from romanized telugu to native telugu script
telugu_transliteration_model.keras: Telugu Translation model in keras format
telugu_transliteration_model.h5: Telugu Translation model in HDF5  format

### CODE REVIEW

File: hindi_transliteration.h5
Comment (arhaankk) [inline hindi_transliteration.h5:1]: This should go in .gitignore

File: classifier.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+df_bn['text'] = df_bn['roman'].apply(clean_text)
+# --- Combine the datasets ---
+data = pd.concat([df_hi[['text', 'language']], 
+                  df_mr[['text', 'language']], 
+                  df_te[['text', 'language']],
+                  df_ml[['text', 'language']],
+                  df_bn[['text', 'language']]],
+                 ignore_index=True)
+
+# print("Combined data sample:")
```
Comment (arhaankk) [inline classifier.py]: remove this code if not needed.

File: classifier.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+max_seq_length = max(len(seq) for seq in sequences)
+X = pad_sequences(sequences, maxlen=max_seq_length, padding='post')
+
+# --- Encode language labels ---
+label_encoder = LabelEncoder()
+y = label_encoder.fit_transform(data['language'])
+print("Languages found:", label_encoder.classes_)
+
+# --- Split the data ---
+X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.1, random_state=42)
```
Comment (arhaankk) [inline classifier.py]: clarification on why `test_size = 0.1` and `random_state = 42`. 

Is it 10% of data is for testing and 90% of data is for training. We can maybe try train on different ratios such as 40-60 split or 30-70 split.

I need the purpose of `random_state =42`

File: classifier.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+# --- Define the classifier model ---
+embedding_dim = 32
```
Comment (arhaankk) [inline classifier.py]: I need clarification on why 32? is it dependent on the neural network?

Have you tried different values. What is the purpose of `embedding_dim = 32`?

File: classifier.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+hidden_units = 64
```
Comment (arhaankk) [inline classifier.py]: This is for the neural network? 64 units in the hidden layer?

File: classifier.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+num_classes = len(label_encoder.classes_)
+
+model = Sequential([
+    Embedding(input_dim=vocab_size, output_dim=embedding_dim, input_shape=(max_seq_length,)),
+    LSTM(hidden_units),
+    Dropout(0.5),
+    Dense(num_classes, activation='softmax')
+])
+
+model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
```
Comment (arhaankk) [inline classifier.py:76]: I see you you are using the sparse categorical cross entropy loss function. Have you tried other loss functions? Like log loss function? or cross entropy loss function?

File: classifier.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+model.summary()
+
+# --- Set up callbacks and train the classifier ---
+callbacks = [
+    EarlyStopping(monitor='val_loss', patience=3, restore_best_weights=True),
```
Comment (arhaankk) [inline classifier.py]: Why is `patience=3`? have you tired other values for patience?

File: databn.tsv
Comment (arhaankk) [inline databn.tsv:1]: Add this to .gitignore

File: datahi.tsv
Comment (arhaankk) [inline datahi.tsv:1]: Add this to .gitignore

File: dataml.tsv
Comment (arhaankk) [inline dataml.tsv:1]: Add this to .gitignore

File: datamr.tsv
Comment (arhaankk) [inline datamr.tsv:1]: Add this to .gitignore

File: datate.tsv
Comment (arhaankk) [inline datate.tsv:1]: Add this to .gitignore

File: hindi_transliteration_model.keras
Comment (arhaankk) [inline hindi_transliteration_model.keras:1]: Add this to .gitignore

File: langhi.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+attn_output = Attention(name='attention')([decoder_lstm, encoder_dense])
+decoder_concat = Concatenate(name='concat')([decoder_lstm, attn_output])
+decoder_output = TimeDistributed(Dense(len(target_tokenizer.word_index) + 1, activation='softmax'),
+                                   name='time_distributed')(decoder_concat)
+
+model = Model(encoder_input, decoder_output)
+model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
+model.summary()
+
+# # 10. Set up callbacks and train the model
```
Comment (arhaankk) [inline langhi.py]: You can remove this maybe?

File: langhi.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    'प', 'फ', 'ब', 'भ', 'म', 'य', 'र', 'ल', 'व', 'श', 'ष', 'स', 'ह',
+    # Matras / vowel modifiers and other signs
+    'ा', 'ि', 'ी', 'ु', 'ू', 'े', 'ै', 'ो', 'ौ', 'ं', 'ः', '्', ' '
+]
+target_tokenizer = Tokenizer(char_level=True, filters='')
+target_tokenizer.fit_on_texts(hindi_tokens)
+
+# 3. Load Hindi dataset from Hugging Face (romanized Hindi to native Hindi)
+# (Make sure you have internet access and the dataset is available.)
+from datasets import load_dataset
```
Comment (arhaankk) [inline langhi.py]: isn't this the same as line 12

File: telugu_transliteration.h5
Comment (arhaankk) [inline telugu_transliteration.h5:1]: add this to .gitignore file

File: telugu_transliteration_model.keras
Comment (arhaankk) [inline telugu_transliteration_model.keras:1]: add this to .gitignore file

File: preprocess.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
+# import pandas as pd
+
+# def process_hindi_tsv(file_path):
```
Comment (arhaankk) [inline preprocess.py]: Remove all unnecessary comments.

File: language_classifier.keras
Comment (arhaankk) [inline language_classifier.keras:1]: add this to .gitignore

File: language_classifier.h5
Comment (arhaankk) [inline language_classifier.h5:1]: Add this to .gitignore

File: langte.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+# 7. Prepare input and target sequences
+source_sequences = source_tokenizer.texts_to_sequences(data['en'])
+X = pad_sequences(source_sequences, maxlen=max_seq_length, padding='post')
+
+target_sequences = target_tokenizer.texts_to_sequences(data['te'])
+y = pad_sequences(target_sequences, maxlen=max_seq_length, padding='post')
+
+# 8. Train-validation split
+X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.1, random_state=42)
```
Comment (arhaankk) [inline langte.py]: Again, why `random_state = 42`?

File: langte.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+decoder_input = RepeatVector(int(max_seq_length), name='decoder_repeat')(encoder_dense[:, -1, :])
+decoder_lstm = LSTM(hidden_units, return_sequences=True, name='decoder_lstm')(decoder_input)
+
+attn_output = Attention(name='attention')([decoder_lstm, encoder_dense])
+decoder_concat = Concatenate(name='concat')([decoder_lstm, attn_output])
+decoder_output = TimeDistributed(Dense(len(target_tokenizer.word_index) + 1, activation='softmax'),
+                                   name='time_distributed')(decoder_concat)
+
+model = Model(encoder_input, decoder_output)
+model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
```
Comment (arhaankk) [inline langte.py]: We can definately loo at other loss functions.

File: langte.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+model.summary()
+
+# 10. Set up callbacks and train the model
```
Comment (arhaankk) [inline langte.py]: You can delete this code, if not needed

**Review by arhaankk — COMMENTED** [review]
@Appearls can you please fix and go though all the reviews.

File: hindi_transliteration.h5
Comment (arhaankk) [reply hindi_transliteration.h5:1]: Add this to a new folder

File: classifier.py (hunk 2 above)
Comment (Appearls) [reply classifier.py]: Purpose of random_state = 42, ensures that the same data ends up in the training set. Changing the random_state can lead to changes in the training data.
Yep, currently since the model is being created with this code I used 90%-10%. We can change it to different ratios for testing

File: classifier.py (hunk 3 above)
Comment (arhaankk) [reply classifier.py]: 32 is the set of number of ways to represent each word.

File: language_classifier.keras
Comment (arhaankk) [reply language_classifier.keras:1]: Add this to a folder

File: language_classifier.h5
Comment (arhaankk) [reply language_classifier.h5:1]: Add this to a folder

File: hindi_transliteration.h5
Comment (Appearls) [reply hindi_transliteration.h5:1]: Added to folder models

File: hindi_transliteration_model.keras
Comment (Appearls) [reply hindi_transliteration_model.keras:1]: Added to folder models

File: telugu_transliteration.h5
Comment (Appearls) [reply telugu_transliteration.h5:1]: Added to folder models

File: telugu_transliteration_model.keras
Comment (Appearls) [reply telugu_transliteration_model.keras:1]: Added to folder models

File: databn.tsv
Comment (Appearls) [reply databn.tsv:1]: Added to data folder

File: dataml.tsv
Comment (Appearls) [reply dataml.tsv:1]: Added to data folder

File: datahi.tsv
Comment (Appearls) [reply datahi.tsv:1]: Added to data folder

File: datamr.tsv
Comment (Appearls) [reply datamr.tsv:1]: Added to data folder

File: datate.tsv
Comment (Appearls) [reply datate.tsv:1]: Added to data folder

File: classifier.py (hunk 5 above)
Comment (Appearls) [reply classifier.py:76]: Currently No. I will different loss functions once the basic system is set up

File: classifier.py (hunk 1 above)
Comment (Appearls) [reply classifier.py]: Done.

File: classifier.py (hunk 6 above)
Comment (Appearls) [reply classifier.py]: Patience is 3, to allow 3 epochs to run even without improvement in validation loss.
5 was taking a bit of time and 2 was less, so I choose 3.

File: langhi.py (hunk 7 above)
Comment (Appearls) [reply langhi.py]: model.summary() provides a detailed overview of model's architecture before training begins. It helps identify setup issues or configuration errors early in the development process.

File: langhi.py (hunk 8 above)
Comment (Appearls) [reply langhi.py]: done!

File: preprocess.py (hunk 9 above)
Comment (Appearls) [reply preprocess.py]: done

File: langte.py (hunk 10 above)
Comment (Appearls) [reply langte.py]: Purpose of random_state = 42, ensures that the same data ends up in the training set. Changing the random_state can lead to changes in the training data.

**Review by arhaankk — COMMENTED** [review]
@Appearls Well Done, you have done a great job. Amazing code.
